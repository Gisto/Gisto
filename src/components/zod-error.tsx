import { BadgeAlert } from 'lucide-react';
import { motion } from 'motion/react';
import { z } from 'zod';

import { formatZodErrors } from '@/lib/utils';

export const ZodError = ({ errors, path }: { errors: z.ZodIssue[]; path: string }) => {
  const error = formatZodErrors(errors)[path];

  if (error) {
    return (
      <motion.div
        className="text-danger text-xs flex items-center gap-2"
        initial={{ opacity: 0, scale: 0.8 }}
        animate={{ opacity: 1, scale: 1 }}
      >
        <BadgeAlert className="stroke-danger size-4" /> {error}
      </motion.div>
    );
  }

  return null;
};
