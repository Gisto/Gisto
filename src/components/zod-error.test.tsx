import { render } from '@testing-library/react';
import { z } from 'zod';

import { ZodError } from './zod-error';

describe('ZodError Component', () => {
  it('renders the error message when an error exists for the given path', () => {
    const errors: z.core.$ZodIssue[] = [
      { path: ['field1'], message: 'Field1 is required', code: 'custom', input: undefined },
    ];
    const path = 'field1';

    const { getByText } = render(<ZodError errors={errors} path={path} />);

    expect(getByText('Field1 is required')).toBeInTheDocument();
  });

  it('renders nothing when no error exists for the given path', () => {
    const errors: z.core.$ZodIssue[] = [
      { path: ['field1'], message: 'Field1 is required', code: 'custom', input: undefined },
    ];
    const path = 'field2';

    const { container } = render(<ZodError errors={errors} path={path} />);

    expect(container.firstChild).toBeNull();
  });

  it('renders with correct styles and animation', () => {
    const errors: z.core.$ZodIssue[] = [
      { path: ['field1'], message: 'Field1 is required', code: 'custom', input: undefined },
    ];
    const path = 'field1';

    const { getByText } = render(<ZodError errors={errors} path={path} />);
    const errorElement = getByText('Field1 is required');

    expect(errorElement).toHaveClass('text-danger text-xs flex items-center gap-2');
  });

  it('matches the snapshot', () => {
    const errors: z.core.$ZodIssue[] = [
      { path: ['field1'], message: 'Field1 is required', code: 'custom', input: undefined },
    ];
    const path = 'field1';

    const { asFragment } = render(<ZodError errors={errors} path={path} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
