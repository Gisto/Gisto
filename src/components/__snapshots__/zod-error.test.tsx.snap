// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ZodError Component > matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="text-danger text-xs flex items-center gap-2"
    style="opacity: 0; transform: scale(0.8);"
  >
    <svg
      class="lucide lucide-badge-alert stroke-danger size-4"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3.85 8.62a4 4 0 0 1 4.78-4.77 4 4 0 0 1 6.74 0 4 4 0 0 1 4.78 4.78 4 4 0 0 1 0 6.74 4 4 0 0 1-4.77 4.78 4 4 0 0 1-6.75 0 4 4 0 0 1-4.78-4.77 4 4 0 0 1 0-6.76Z"
      />
      <line
        x1="12"
        x2="12"
        y1="8"
        y2="12"
      />
      <line
        x1="12"
        x2="12.01"
        y1="16"
        y2="16"
      />
    </svg>
     Field1 is required
  </div>
</DocumentFragment>
`;
